### DISCUSSION TITLE
Usage of gotd for e2e testing

### DISCUSSION THREAD

**Discussion by 9seconds** [gotd/td#286] [Q&A] [answered]
Hey! Thanks for the project!

I have tried to use gotd for testing of https://github.com/9seconds/mtg but can't figure out if it is possible to do that. Let me show you an example: https://github.com/9seconds/mtg/blob/master/mtglib/proxy_test.go#L139-L169

This is a rather simple test. I start mtg and send HTTPS requests to this proxy. My goal is to see a response from the fronting domain. I wanna do the same test but for Telegram itself:

1. Start mtg
2. Configure gotd to use this proxy (it uses faketls only if it matters)
3. Send Ping request
4. Get Pong back.

AFAIK, Ping and Pongs are authorization-free and can be sent without any credentials.

How can I make such a test? I do not need anything smart here. I just want to get a response from Telegram to check e2e connectivity.

**Comment by ghost** [comment, marked as answer]
Hi!

There is example how to use MTProxy transport 
https://pkg.go.dev/github.com/gotd/td@v0.33.3/telegram#example-Client-Mtproxy
```go
        ...
	resolver, err := dcs.MTProxyResolver(
		os.Getenv("PROXY_ADDR"),
		secret,
		dcs.MTProxyOptions{},
	)
	if err != nil {
		return xerrors.Errorf("create MTProxy resolver: %w", err)
	}
       ...
```

Also, you can use your own [`DialFunc`](https://pkg.go.dev/github.com/gotd/td@v0.33.3/telegram/dcs#DialFunc) and pass it to the `MTProxyResolver` in [`MTProxyOptions`](https://pkg.go.dev/github.com/gotd/td@v0.33.3/telegram/dcs#MTProxyOptions) 

It can be useful for testing transport through [`net.Pipe`](https://pkg.go.dev/net#Pipe).

**Reply by ernado** [reply]
@9seconds 
Do you want to use telegram testing servers or you want embedded telegram server for e2e tests?
Internally wee are using very basic telegram server in go for our local e2e tests in some cases.

**Reply by 9seconds** [reply]
I would like to use Telegram servers. This gonna be a simple integration test

**Comment by 9seconds** [comment]
Thanks, guys! https://github.com/9seconds/mtg/pull/184
